From bc76d68e3cb2ad6f9da7f6b50316d9312c0e4598 Mon Sep 17 00:00:00 2001
From: Ido Reis <idox.reis@intel.com>
Date: Tue, 10 Jan 2017 19:58:16 +0200
Subject: [PATCH 1/2] mac80211: add net_device_to_ieee80211_vif

extract the vif object from net_device instance in order to bypass
the mac80211 layer. this is done to be able to replace the data
path callbacks in mac80211.

Signed-off-by: Ido Reis <idox.reis@intel.com>
---
 include/net/mac80211.h |  9 +++++++++
 net/mac80211/util.c    | 12 ++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index c0d41fc..800e0cb8 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1501,6 +1501,15 @@ struct ieee80211_vif *wdev_to_ieee80211_vif(struct wireless_dev *wdev);
 struct wireless_dev *ieee80211_vif_to_wdev(struct ieee80211_vif *vif);
 
 /**
+ * net_device_to_ieee80211_vif - return a vif from net_device
+ * @dev: the net_device to get the vif for
+ *
+ * This can be used by mac80211 drivers which needs to manipulate data
+ * path with tier own implementation of net_device_ops.
+ */
+struct ieee80211_vif *net_device_to_ieee80211_vif(struct net_device *dev);
+
+/**
  * ieee80211_vif_to_name - return the vif name
  * @vif: the vif to get the wdev for
  *
diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index d2e3e7f..7481bc8c 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -723,6 +723,18 @@ struct wireless_dev *ieee80211_vif_to_wdev(struct ieee80211_vif *vif)
 }
 EXPORT_SYMBOL_GPL(ieee80211_vif_to_wdev);
 
+struct ieee80211_vif *net_device_to_ieee80211_vif(struct net_device *dev)
+{
+	struct ieee80211_sub_if_data *sdata;
+
+	if (!dev)
+		return NULL;
+
+	sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+	return &sdata->vif;
+}
+EXPORT_SYMBOL_GPL(net_device_to_ieee80211_vif);
+
 const char *ieee80211_vif_to_name(struct ieee80211_vif *vif)
 {
 	struct ieee80211_sub_if_data *sdata;
-- 
2.8.4.windows.1

